Framework/PythonInterface/mantid/utils/dgs/_geometry.py

```diff
@@ -0,0 +1,299 @@
+        # check if float
+        DeltaE = float(DeltaE)
+        DeltaE = np.full(len_hkl, DeltaE)
+    except ValueError:
+        raise ValueError('DeltaE is not a float or numpy array')
+    except TypeError:
+        if len(DeltaE) != len_hkl:
+            raise ValueError('DeltaE has different length than hkl')
+
+    try:
+        # check if int
+        sign = int(sign)
+        sign = np.full(len_hkl,sign)
+    except ValueError:
+        raise ValueError('sign is not an int or numpy array')
+    except TypeError:
+        if len(sign) != len_hkl:
+            raise ValueError('sign has different length than hkl')
+
+    try:
+        UB = lattice.getUB() * 2. * np.pi
+    except:
+        raise ValueError("Can't get the UB matrix from the lattice object")
+
+    # inputs for geometry and goniometer constraints
+    if detector_constraints:
+        if horizontal_extent[0]<-180 or horizontal_extent[1]<horizontal_extent[0] or horizontal_extent[1]>180:
+            raise ValueError("Horizontal constraints must obey -180 <= horizontal_extent[0] <= "
+                             "horizontal_extent[1] <=180")
+        if vertical_extent[0]<-180 or vertical_extent[1]<vertical_extent[0] or vertical_extent[1]>180:
+            raise ValueError("Vertical constraints must obey -180 <= vertical_extent[0] <= "
+                             "vertical_extent[1] <=180")
+        if horizontal_extent_low[0]<-180 or horizontal_extent_low[1]<horizontal_extent_low[0] or horizontal_extent_low[1]>180:
+            raise ValueError("Horizontal constraints must obey -180 <= horizontal_extent_low[0] <= "
+                             "horizontal_extent_low[1] <=180")
+        if vertical_extent_low[0]<-180 or vertical_extent_low[1]<vertical_extent_low[0] or vertical_extent_low[1]>180:
+            raise ValueError("Vertical constraints must obey -180 <= vertical_extent_low[0] <= "
+                             "vertical_extent_low[1] <=180")
+
+    if goniometer_constraints:
+        if goniometer_range[1]<goniometer_range[0] or \
+           goniometer_range[0]<-180. or goniometer_range[1]>180.:
+            raise ValueError("goniometer_range must be an increasing array, "
+                             "with both limits between -180 and 180 degrees")
+
+
+    return (Ei, DeltaE, sign, UB)
+
+
+
+@namedtuplefy
+def qangle(*,  # force keyword arguments
+           Ei: float or np.array,
+           hkl: np.array,  #numpy array of triplets
+           DeltaE: float or np.array,
+           sign: int or np.array,
+           lattice: OrientedLattice,
+           detector_constraints: bool = False,
+           horizontal_extent: np.array = np.array([-180.,180.]),
+           vertical_extent: np.array = np.array([-90.,90.]),
+           horizontal_extent_low: np.array = np.array([0.,0.]),
+           vertical_extent_low: np.array = np.array([0.,0.]),
+           goniometer_constraints: bool = False,
+           goniometer_range: np.array = np.array([-180.,180.]),
```

Comment:
Not a point to change in this PR, but I'd recommend tying things up into classes only used in this file (a bit like C/C++ structs) in the future:

```
from mantid.py36compat import dataclass
@dataclass
class _Extents:
           horizontal_extent: np.array = np.array([-180.,180.]),
           vertical_extent: np.array = np.array([-90.,90.]),
           horizontal_extent_low: np.array = np.array([0.,0.]),
           vertical_extent_low: np.array = np.array([0.,0.]),
```

```
def qangle(..., detector_constraints: bool = False, extents = _Extents(), ...):
    my_val = extents.vertical_extent
```

This helps future developers group things together mentally too, and reduces the number of variables floating around at any given point

Reply:
My reply to @wdzhou 's similar question `Yes, but too complicated for user scripts. I am thinking about keeping it simple, so I would not change it` The function is intended for instrument scientists/users, not for developers. Adding extra classes will just lead to confusion.